lifetimes/fitters/pareto_nbd_fitter.py

```diff
@@ -525,19 +528,64 @@ def _fit(
             if minimize_options["disp"]:
                 print("Optimize function with {}".format(fit_method))
 
-            output = minimize(
-                self._negative_log_likelihood,
-                method=fit_method,
-                tol=tol,
-                x0=current_init_params,
-                args=minimizing_function_args,
-                options=minimize_options,
-            )
+            # Convert printed numpy warnings to actual RuntimeWarnings
+            numpy.seterr('warn')
+            # Use python's warnings module to catch warnings as if they're errors,
+            # specifically warnings with an 'invalid value encountered in subtract'
+            # message, which is the warning that implies an infinite or nan
+            # log-likelihood value
+            warnings.filterwarnings('error', message='invalid value encountered in subtract')
+
+            stored_params = []
+            def print_and_store_params(xk):
+                print(xk)
+                stored_params.append(xk)
+
+            try:
+                output = minimize(
+                    self._negative_log_likelihood,
+                    method=fit_method,
+                    tol=tol,
+                    x0=current_init_params,
+                    callback=print_and_store_params,
+                    args=minimizing_function_args,
+                    options=minimize_options,
+                )
+            except RuntimeWarning:
+                print("Warning encountered, retrying...")
+                last_params = stored_params[-1]
+                f, r, t, weights, reg_param = minimizing_function_args
+                likelihoods = ParetoNBDFitter._conditional_log_likelihood(last_params, f, r, t)
+
+                inf_indices = np.where(np.isinf(likelihoods))
+                nan_indices = np.where(np.isnan(likelihoods))
+                bad_indices = np.concatenate((inf_indices[0], nan_indices[0]))
+                print(f"Number of bad indices: {bad_indices.size}")
+                print(f"Problematic entries: {f[bad_indices]}, {r[bad_indices]}, {t[bad_indices]}")
+
+                f = np.delete(f, bad_indices)
+                r = np.delete(r, bad_indices)
+                t = np.delete(t, bad_indices)
+                weights = np.delete(weights, bad_indices)
+
+                minimizing_function_args = (f, r, t, weights, reg_param)
+
+                output = minimize(
```

Comment:
Should we make this a recursive call to `_fit` so we have the option to recover more than once?

Reply:
This isn't a traditional retry mechanism, in that once we've removed the violating rows we should run the optimization once, but I don't think there's any benefit in attempting to retry again

Reply:
So if we remove the rows, retry, and run into nan/infs again we'll fail. I'm not sure if this is the right mental model, but I was picturing a dataset with 2 outliers where one is much more of an outlier than the other. It seems reasonable that the large outlier would generate a nan, get removed, but then after a few iterations the smaller outlier would also over/under flow.

Do you think that's unlikely to happen? Or that if we get in a situation like that we'd want to fail and investigate? Or that it's just not worth the level of effort to handle multiple retries at the moment? For context, I'm fine if any of these is the case, just trying to get a good understanding of the problem.

Reply:
Ah got it. It's a good thought. I do agree it seems that theoretically the params could move in a direction that causes some other row to throw nans/infs, though I haven't seen it in practice. I can make this more robust tho

Reply:
Ok cool, glad we're on the same page.

Using a recursive call seemed like a cheap way to make this more robust, but if there's more complexity than that it might not be worth it until we have evidence we need it.